Web Inspector: Canvas: expand OffscreenCanvas support

dcrousso opened the pull request:
#### 78a437817ab22672737ec12685469f4bc84275f3
<pre>
Web Inspector: Canvas: expand OffscreenCanvas support
<a href="https://bugs.webkit.org/show_bug.cgi?id=256747">https://bugs.webkit.org/show_bug.cgi?id=256747</a>

Reviewed by Patrick Angle.

`OffscreenCanvas` currently supports more than just `&quot;2d&quot;`, so Web Inspector should too.

* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::releaseObjectForRecording):
Add `offscreen-*` variants for `bitmaprenderer`, `webgl`, and `webgl2`.
Drive-by: Fix an issue that prevented getting the content of an `OffscreenCanvas`.

* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas.fromPayload):
(WI.Canvas.displayNameForContextType):
(WI.Canvas.displayNameForColorSpace):
(WI.Canvas.prototype.get supportsRecording): Added.
(WI.Canvas.prototype.get is2D): Added.
(WI.Canvas.prototype.get isBitmapRender): Added.
(WI.Canvas.prototype.get isWebGL): Added.
(WI.Canvas.prototype.get isWebGL2): Added.
(WI.Canvas.prototype.is2D): Deleted.
* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording):
(WI.Recording.fromPayload):
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.get isCanvas): Added.
(WI.Recording.prototype.get isCanvas2D): Added.
(WI.Recording.prototype.get isCanvasBitmapRender): Added.
(WI.Recording.prototype.get isCanvasWebGL): Added.
(WI.Recording.prototype.get isCanvasWebGL2): Added.
(WI.Recording.prototype.createContext):
(WI.Recording.prototype.async _process):
(WI.Recording.is2D): Deleted.
Add helpers for identifying the context type regardless of whether it&apos;s from `HTMLCanvasElement` or `OffscreenCanvas` (since the API of the rendering context is (almost) identical in most cases).

* Source/WebInspectorUI/UserInterface/Models/RecordingAction.js:
(WI.RecordingAction._prototypeForType):
(WI.RecordingAction.prototype.process):
(WI.RecordingAction.prototype.process.getContent):
(WI.RecordingAction.prototype.async swizzle):
* Source/WebInspectorUI/UserInterface/Models/RecordingState.js:
(WI.RecordingState.prototype.fromCanvasContext2D): Renamed from `fromContext`.
(WI.RecordingState.async swizzleInitialState):
* Source/WebInspectorUI/UserInterface/Models/ShaderProgram.js:
(WI.ShaderProgram):
(WI.ShaderProgram.contextTypeSupportsProgramType):
(WI.ShaderProgram.prototype.get displayName):
(WI.ShaderProgram.prototype.set disabled):
(WI.ShaderProgram.prototype.showHighlight):
(WI.ShaderProgram.prototype.hideHighlight):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/CanvasSidebarPanel.js:
(WI.CanvasSidebarPanel.prototype._canvasChanged):
(WI.CanvasSidebarPanel.prototype._updateRecordNavigationItem):
* Source/WebInspectorUI/UserInterface/Views/RecordingContentView.js:
(WI.RecordingContentView):
(WI.RecordingContentView.prototype.get navigationItems):
(WI.RecordingContentView.prototype.attached):
(WI.RecordingContentView.prototype._export):
(WI.RecordingContentView.prototype._updateExportButton):
(WI.RecordingContentView.prototype._handleExportNavigationItemClicked):
* Source/WebInspectorUI/UserInterface/Views/RecordingStateDetailsSidebarPanel.js:
(WI.RecordingStateDetailsSidebarPanel.prototype.inspect):
(WI.RecordingStateDetailsSidebarPanel.prototype.set action):
* Source/WebInspectorUI/UserInterface/Views/ShaderProgramContentView.js:
(WI.ShaderProgramContentView.prototype.get saveData):
* Source/WebInspectorUI/UserInterface/Views/ShaderProgramTreeElement.js:
(WI.ShaderProgramTreeElement):
(WI.ShaderProgramTreeElement.prototype.onattach):
(WI.ShaderProgramTreeElement.prototype.ondetach):
(WI.ShaderProgramTreeElement.prototype.populateContextMenu):
Use the new helpers instead of directly checking the `contextType`/`type` where able.

* Source/WebCore/inspector/InspectorFrontendHost.idl:
* Source/WebCore/inspector/InspectorFrontendHost.h:
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::getCurrentX const):
(WebCore::InspectorFrontendHost::getCurrentY const):
(WebCore::InspectorFrontendHost::getPath const):
(WebCore::InspectorFrontendHost::setPath const):
* Source/WebInspectorUI/UserInterface/Base/IDLExtensions.js:
Add IDL extensions for `OffscreenCanvasRenderingContext2D`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/canvas/resources/recording-2d.js:
(async load):
* LayoutTests/inspector/canvas/resources/recording-bitmaprenderer.js:
(load):
* LayoutTests/inspector/canvas/resources/recording-webgl.js:
(load):
* LayoutTests/inspector/canvas/resources/recording-webgl2.js:
(load):
* LayoutTests/inspector/canvas/resources/shaderProgram-utilities-webgl.js:
(createProgram):
* LayoutTests/inspector/canvas/console-record-offscreen-bitmaprenderer.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-bitmaprenderer-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d.html:
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d-expected.txt:
* LayoutTests/inspector/canvas/console-record-offscreen-webgl.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl2.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl2-expected.txt: Added.
* LayoutTests/inspector/canvas/create-context-bitmaprenderer.html:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl.html:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2.html:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-frameCount-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-full-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-memoryLimit-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-snapshots.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-snapshots-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-snapshots-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-frameCount-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-full-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-memoryLimit-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-snapshots.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-snapshots-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-snapshots-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/271300@main">https://commits.webkit.org/271300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f78b59e53c354e9efb4dd3d02703f45cb586e3a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30459 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25485 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8556 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3961 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28194 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23995 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4589 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4775 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31148 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/24202 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/simple-inline-stacktrace-with-catch.js.wasm-eager (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25443 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31027 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27078 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4782 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2941 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28840 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24761 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/34557 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6707 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5228 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7476 "Found 4078 new JSC stress test failures: stress/sampling-profiler-display-name.js.dfg-eager, wasm.yaml/wasm/branch-hints/branchHintsSection.js.default-wasm, wasm.yaml/wasm/branch-hints/branchHintsSection.js.wasm-collect-continuously, wasm.yaml/wasm/branch-hints/branchHintsSection.js.wasm-eager, wasm.yaml/wasm/branch-hints/branchHintsSection.js.wasm-eager-jettison, wasm.yaml/wasm/branch-hints/branchHintsSection.js.wasm-no-cjit, wasm.yaml/wasm/branch-hints/branchHintsSection.js.wasm-slow-memory, wasm.yaml/wasm/extended-const-spec-tests/data.wast.js.default-wasm, wasm.yaml/wasm/extended-const-spec-tests/data.wast.js.wasm-collect-continuously, wasm.yaml/wasm/extended-const-spec-tests/data.wast.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5255 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Canvas.json:
```diff
@@ -23,7 +23,12 @@
             "type": "string",
-            "enum": ["canvas-2d", "offscreen-canvas-2d", "bitmaprenderer", "webgl", "webgl2"],
+            "enum": [
+                "canvas-2d", "offscreen-canvas-2d",
```
Makes me wonder if "offscreen" should just be another boolean instead of a context type? (Better to do so know when we only have to shim compatibility for a single existing value)

dcrousso replied:
id prefer things be explicit like this as it means we can immediately understand what's going on rather than having to look for the source of truth in two places (especially since the API is not always the same between main page and offscreen)

also, if a context type is added in the future that only exists in one then that would be more annoying to express

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Recording.json:
```diff
@@ -8,7 +8,12 @@
             "type": "string",
-            "enum": ["canvas-2d", "offscreen-canvas-2d", "canvas-bitmaprenderer", "canvas-webgl", "canvas-webgl2"],
+            "enum": [
+                "canvas-2d", "offscreen-canvas-2d",
```
Ditto from Canvas.json:27

patrickangle commented on Source/WebCore/inspector/InspectorFrontendHost.cpp:
```diff
@@ -860,4 +861,28 @@ void InspectorFrontendHost::setPath(CanvasRenderingContext2D& context, Path2D& p
 
+#if ENABLE(OFFSCREEN_CANVAS)
+
+float InspectorFrontendHost::getCurrentX(const OffscreenCanvasRenderingContext2D& context) const
```
I'm getting a sense of deja vu, but I still wish these method names were more specific (e.g. `getCanvasCurrentX`). But I know there is a reason we didn't that I've since forgotten.

dcrousso replied:
we definitely coulda done that, but IMO the argument type should be a good enough clue as to what's going on (at least in C++)

i think you're thinking of `Source/WebInspectorUI/UserInterface/Base/IDLExtensions.js`

patrickangle commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -301,8 +300,18 @@ localizedStrings["Cancellation @ Local Override Type"] = "Cancellation";
 localizedStrings["Canvas %s"] = "Canvas %s";
-/* Bitmap Renderer is a type of rendering context associated with a <canvas> element */
+/* 2D is a type of rendering context associated with a <canvas> element. */
+localizedStrings["Canvas Context Type 2D"] = "2D";
```
We aren't super consistent about this, but should this be `2D @ Canvas Context Type` (same below)?

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/Canvas.js:
```diff
@@ -147,9 +162,9 @@ WI.Canvas = class Canvas extends WI.Object
         switch(colorSpace) {
         case WI.Canvas.ColorSpace.SRGB:
-            return WI.unlocalizedString("sRGB");
+            return WI.UIString("sRGB", "sRGB @ Color Space", "Label for a canvas that uses the sRGB color space.");
```
Not sure it's correct that we begin translating these color space names. To my knowledge sRGB and Display P3 aren't typically localized. For example, the product page for Apple Studio Display in France says `Internet et Web (sRGB)` and in the US says `Internet and Web (sRGB)`.

dcrousso replied:
id rather leave that up to the localizers, as that may not always be true in every language (e.g. Japanese might prefer to use katakana instead of english)

my approach is usually "unless what we're showing is a code snippet or something from the developer, everything should have the opportunity to be localized"

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/Canvas.js:
```diff
@@ -215,11 +252,26 @@ WI.Canvas = class Canvas extends WI.Object
     }
 
-    is2D()
+    get is2D()
```
For consistency with the `Recording` class could this be `isCanvas2D` (same for WebGL ones below)?

dcrousso replied:
IMO that's a bit redundant since this is a `WI.Canvas` object (e.g. id find it weird to read `canvas.isCanvas2D`)

FWIW the only reason `WI.Recording` uses `isCanvas2D` instead of just `is2D` is because there's technically nothing about `WI.Recording` that's specific to `<canvas>`.  when i first built it many years ago my thinking was that we'd be able to reuse the general purpose data structure for other kinds of recordings (e.g. JS object mutations, DOM tree shenanigans, etc.)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ShaderProgramTreeElement.js:
```diff
@@ -33,8 +33,7 @@ WI.ShaderProgramTreeElement = class ShaderProgramTreeElement extends WI.GeneralT
         super("shader-program", shaderProgram.displayName, subtitle, shaderProgram);
 
         // FIXME: add support for disabling/highlighting WebGPU shader pipelines.
```
I wonder if there is a bug tracking this, or if this is the most aspirational type of `FIXME` (the type no one filed)?

dcrousso replied:
looks like there is <[https://webkit.org/b/201856]()>

not sure why it isn't linked here 🤔

dcrousso committed:
Web Inspector: Canvas: expand OffscreenCanvas support
https://bugs.webkit.org/show_bug.cgi?id=256747

Reviewed by Patrick Angle.

`OffscreenCanvas` currently supports more than just `"2d"`, so Web Inspector should too.

* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::releaseObjectForRecording):
Add `offscreen-*` variants for `bitmaprenderer`, `webgl`, and `webgl2`.
Drive-by: Fix an issue that prevented getting the content of an `OffscreenCanvas`.

* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas.fromPayload):
(WI.Canvas.displayNameForContextType):
(WI.Canvas.displayNameForColorSpace):
(WI.Canvas.prototype.get supportsRecording): Added.
(WI.Canvas.prototype.get is2D): Added.
(WI.Canvas.prototype.get isBitmapRender): Added.
(WI.Canvas.prototype.get isWebGL): Added.
(WI.Canvas.prototype.get isWebGL2): Added.
(WI.Canvas.prototype.is2D): Deleted.
* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording):
(WI.Recording.fromPayload):
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.get isCanvas): Added.
(WI.Recording.prototype.get isCanvas2D): Added.
(WI.Recording.prototype.get isCanvasBitmapRender): Added.
(WI.Recording.prototype.get isCanvasWebGL): Added.
(WI.Recording.prototype.get isCanvasWebGL2): Added.
(WI.Recording.prototype.createContext):
(WI.Recording.prototype.async _process):
(WI.Recording.is2D): Deleted.
Add helpers for identifying the context type regardless of whether it's from `HTMLCanvasElement` or `OffscreenCanvas` (since the API of the rendering context is (almost) identical in most cases).

* Source/WebInspectorUI/UserInterface/Models/RecordingAction.js:
(WI.RecordingAction._prototypeForType):
(WI.RecordingAction.prototype.process):
(WI.RecordingAction.prototype.process.getContent):
(WI.RecordingAction.prototype.async swizzle):
* Source/WebInspectorUI/UserInterface/Models/RecordingState.js:
(WI.RecordingState.prototype.fromCanvasContext2D): Renamed from `fromContext`.
(WI.RecordingState.async swizzleInitialState):
* Source/WebInspectorUI/UserInterface/Models/ShaderProgram.js:
(WI.ShaderProgram):
(WI.ShaderProgram.contextTypeSupportsProgramType):
(WI.ShaderProgram.prototype.get displayName):
(WI.ShaderProgram.prototype.set disabled):
(WI.ShaderProgram.prototype.showHighlight):
(WI.ShaderProgram.prototype.hideHighlight):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/CanvasSidebarPanel.js:
(WI.CanvasSidebarPanel.prototype._canvasChanged):
(WI.CanvasSidebarPanel.prototype._updateRecordNavigationItem):
* Source/WebInspectorUI/UserInterface/Views/RecordingContentView.js:
(WI.RecordingContentView):
(WI.RecordingContentView.prototype.get navigationItems):
(WI.RecordingContentView.prototype.attached):
(WI.RecordingContentView.prototype._export):
(WI.RecordingContentView.prototype._updateExportButton):
(WI.RecordingContentView.prototype._handleExportNavigationItemClicked):
* Source/WebInspectorUI/UserInterface/Views/RecordingStateDetailsSidebarPanel.js:
(WI.RecordingStateDetailsSidebarPanel.prototype.inspect):
(WI.RecordingStateDetailsSidebarPanel.prototype.set action):
* Source/WebInspectorUI/UserInterface/Views/ShaderProgramContentView.js:
(WI.ShaderProgramContentView.prototype.get saveData):
* Source/WebInspectorUI/UserInterface/Views/ShaderProgramTreeElement.js:
(WI.ShaderProgramTreeElement):
(WI.ShaderProgramTreeElement.prototype.onattach):
(WI.ShaderProgramTreeElement.prototype.ondetach):
(WI.ShaderProgramTreeElement.prototype.populateContextMenu):
Use the new helpers instead of directly checking the `contextType`/`type` where able.

* Source/WebCore/inspector/InspectorFrontendHost.idl:
* Source/WebCore/inspector/InspectorFrontendHost.h:
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::getCurrentX const):
(WebCore::InspectorFrontendHost::getCurrentY const):
(WebCore::InspectorFrontendHost::getPath const):
(WebCore::InspectorFrontendHost::setPath const):
* Source/WebInspectorUI/UserInterface/Base/IDLExtensions.js:
Add IDL extensions for `OffscreenCanvasRenderingContext2D`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/canvas/resources/recording-2d.js:
(async load):
* LayoutTests/inspector/canvas/resources/recording-bitmaprenderer.js:
(load):
* LayoutTests/inspector/canvas/resources/recording-webgl.js:
(load):
* LayoutTests/inspector/canvas/resources/recording-webgl2.js:
(load):
* LayoutTests/inspector/canvas/resources/shaderProgram-utilities-webgl.js:
(createProgram):
* LayoutTests/inspector/canvas/console-record-offscreen-bitmaprenderer.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-bitmaprenderer-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d.html:
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d-expected.txt:
* LayoutTests/inspector/canvas/console-record-offscreen-webgl.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl2.html: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-webgl2-expected.txt: Added.
* LayoutTests/inspector/canvas/create-context-bitmaprenderer.html:
* LayoutTests/inspector/canvas/create-context-bitmaprenderer-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl.html:
* LayoutTests/inspector/canvas/create-context-webgl-expected.txt:
* LayoutTests/inspector/canvas/create-context-webgl2.html:
* LayoutTests/inspector/canvas/create-context-webgl2-expected.txt:
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-bitmaprenderer-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-frameCount-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-full-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-memoryLimit-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-snapshots.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl-snapshots-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl-snapshots-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-frameCount-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-full-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-memoryLimit-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-snapshots.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-webgl2-snapshots-expected.txt: Added.
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-offscreen-webgl2-snapshots-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Canonical link: https://commits.webkit.org/271300@main
